best-first-cv/script.whizzml

```diff
@@ -0,0 +1,561 @@
+  (let (fields ((fetch dataset-id) "fields"))
+    (map (lambda (id) (fields [id "name"])) ids)))
+
+;; Get the default set of input fields for this dataset (all preferred
+;; fields minus the objective field).
+(define (default-inputs dataset-id obj-id)
+  (let (fields ((fetch dataset-id) "fields")
+        fids (keys fields))
+    (filter (lambda (k) (and (fields [k "preferred"] false) (not (= obj-id k))))
+            fids)))
+
+
+;; Get the set of input fields for this dataset by using the importance of the
+;; fields in a random decision forest
+(define (important-fields dataset-id obj-id)
+  (let (ensemble-id (create-and-wait-ensemble dataset-id
+                                              {"randomize" true
+                                               "objective_field" obj-id})
+        ensemble (fetch ensemble-id)
+        importance (ensemble "importance")
+        importance (for (key (keys importance))
+                     {"feature" key "phi" (importance key)}))
+    (reverse (sort-by-key "phi" importance))))
+
+
+;; Make a list of input fields candidates given the potential
+;; features in potentials.
+(define (make-candidates selected potentials)
+  (let (make-req (lambda (fid)
+                   (cons fid selected)))
+    (map make-req potentials)))
+
+;; Given a set of dataset folds, a list of potential features, and a list of
+;; model ids corresponding to those features, select the best
+;; potential feature by performing an evaluation on each model and
+;; returning the feature with the best performance.
+(define (select-feature k-fold-datasets objective-id options candidates potentials)
+  (let (make-attrs (lambda (inp) (merge {"input_fields" inp}
+                                        options))
+        potentials (map (lambda (x) (x "feature")) potentials)
+        all-reqs (map make-attrs candidates)
+        cvs (map (lambda (e-attrs) (cross-validation k-fold-datasets
+                                                     objective-id
+                                                     e-attrs
+                                                     {}
+                                                     true)) all-reqs)
+        ;;vs (map (lambda (ev) (ev ["result" "model" "average_phi"] 0)) cvs)
+        vs cvs
+        value-map (make-map potentials vs)
+        max-val (get-max vs)
+        choose-best (lambda (id) (and (= max-val (value-map id)) id)))
+    {"feature" (some choose-best potentials)
+     "phi" max-val}))
+
+;; Retrieves and checks whether the objective field of the given
+;; dataset is categorical, raising an error otherwise.
+(define (get-objective ds-id obj-id)
+  (let (obj-id (if (empty? obj-id)
+                   (dataset-get-objective-id ds-id)
+                   obj-id)
+        otype ((fetch ds-id) ["fields" obj-id "optype"] "missing"))
+    (when (not (= "categorical" otype))
+      (raise (str "The dataset's objective field must be categorical, "
+                  "but is " otype)))
```

Comment:
in the code above, every time we raise, it's a map with "message"/"code", but here it's only a string... if there's no special reason, we can homogenize (one way or the other, no strong opinion: since this is a script and nobody's catching the exception, it can be a string everywhere... hmmm, or are you thinking that in the cases above the code will eventually be in a library? if so, it makes sense to throw there a map with code and here a string, so this would be perfectly fine).

Reply:
yes, The part that had code+message was thought to be a library in the future.